### DISCUSSION TITLE
potential race condition in Build.progressive_output

### DISCUSSION THREAD

**Discussion by dd-dino** [joelee2012/api4jenkins#105] [General]
Hello,

I would like to point out to a race condition that my client program hits occasionally. It should be easy to fix.

I find my client in an infinite loop between lines:
https://github.com/joelee2012/api4jenkins/blob/main/api4jenkins/build.py#L26-L30

After doing strace on a hung client process, I found out the following:
- http request has start=213928
- http response has X-Text-Size: 213928
- there is no X-More-Data header in the response

It appears the client had read all the output from the build when the build was not quite done yet. 
Then the client made another request only to find that there is no more output. At that point the client does not check about X-More-Data header and just loops back to make another request. It remain in this infinite loop :/.

Do you accept pull requests? Is this something you could fix?

**Comment by joelee2012** [comment]
Yes, welcome to create PR.

**Reply by dd-dino** [reply]
I don't have write access, couldn't push a branch. (I guess I would have had to fork the repository, etc. Seems too much work for the little change.)
So, here's the diff. Hope you will consider it.
Thanks!
[a4j.txt](https://github.com/user-attachments/files/16339526/a4j.txt)

**Reply by joelee2012** [reply]
@dd-dino i can't reproduce your issue with following pipeline, are you able  to share your pipeline ?
```groovy
pipeline{
   agent {
       any
     }
    stages {
        stage('sleep') {
            steps {
                sh 'sleep 60'
            }
        }
    }
}
```

**Reply by dd-dino** [reply]
My pipeline is rather complex. I have a job that runs a python program that controls other jobs. And that's where it happens.
To reproduce, you can apply the build_test.py changes that add a new unit test. That unit test fails without the changes in build.py.
Hope that helps.
